Allow user to set forbidden headers in XHR

NoneGiven opened the pull request:
Even in WebExtensions, Firefox enforces [per spec](https://fetch.spec.whatwg.org/#forbidden-header-name) that Referer, Origin, and other headers may not be set by the developer on the XHR object. However, if the outgoing XHR is intercepted in the `onBeforeSendHeaders` event, it is possible to set these headers as desired.

~~This is a WIP~~ This change currently looks for Referer and Origin and shows how GM can set the header values when the user supplies them the normal way, through the `headers` property of the object passed to `GM.xmlHttpRequest`. It's necessary for GM to add dummy headers to preserve the user's header values until `onBeforeSendHeaders`, at which point we can set the real header and delete the dummy.

Fixes https://github.com/greasemonkey/greasemonkey/issues/2599

NoneGiven commented:
One thing that needs investigation is whether the `<all_urls>` in the listener registration can hopefully be replaced with `*://*/*.user.js`. We don't actually want GM running `rewriteHeaders` for every single request the browser makes.

xor10 commented:
Use `originUrl` to check if requests are from extension. :)
Should be allowed to modify the `Host` and `User-Agent` headers too.

```javascript
const extensionUrl = chrome.extension.getURL('');
function rewriteHeaders(detail) {
  // rewrite only requests from background script (GM.xmlHttpRequest)
  if (detail.originUrl && detail.originUrl.startsWith(extensionUrl)) {
    const newHeaders = [];
    ...

    return { requestHeaders: newHeaders };
  }

  return { requestHeaders: detail.requestHeaders };
}

```

NoneGiven commented:
Thanks @xor10 ! I added the originUrl check and also rewrote the code to actually work (had some confusion over which objects were plain objects and which were arrays). I tested this with the modified GM code loaded into FF 58 and it works great.

Setting `User-Agent` is already possible. For the other two, maybe @arantius can decide on the final list of forbidden headers to support.

Sxderp commented on src/bg/on-user-script-xhr.js:
```diff
@@ -119,4 +127,37 @@ function open(xhr, d, port) {
+      var prefixedHeader = getHeader(e.requestHeaders, dummyHeaderPrefix + name);
+      if (prefixedHeader) {
+        if (!getHeader(e.requestHeaders, name)) {
+          // only try to add real header if request doesn't already have it
```
I'm not sure I understand the purpose of this conditional, could you provide an example of when `getHeader` may return true on line 147?

Further, under what condition would you _not_ want to replace a header with the prefixed version?

NoneGiven replied:
It would return true for headers like `Accept-Encoding` and `Connection` that are always automatically set, if we were allowing the user to replace those.

>under what condition would you not want to replace a header with the prefixed version?

I guess you wouldn't.

NoneGiven commented:
@arantius Anything I can do to move this PR along? Unlike that other one, this one works and has a legitimate goal (giving `GM.xmlHttpRequest` the same utility as the old `GM_xmlhttpRequest`).

arantius commented:
Sorry!  Merging now.
